File: Go/src/github.com/cadence-proxy/internal/messages/workflow_future_ready_reply.go
//-----------------------------------------------------------------------------
// FILE:		workflow_future_ready_reply.go
// CONTRIBUTOR: John C Burns
// COPYRIGHT:	Copyright (c) 2016-2019 by neonFORGE, LLC.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package messages

import (
	internal "github.com/cadence-proxy/internal"
)

type (

	// WorkflowFutureReadyReply is a WorkflowReply of MessageType
	// WorkflowFutureReadyReply.  It holds a reference to a WorkflowReply in memory
	// and is the reply type to a WorkflowFutureReadyRequest
	WorkflowFutureReadyReply struct {
		*WorkflowReply
	}
)

// NewWorkflowFutureReadyReply is the default constructor for
// a WorkflowFutureReadyReply
//
// returns *WorkflowFutureReadyReply -> a pointer to a newly initialized
// WorkflowFutureReadyReply in memory
func NewWorkflowFutureReadyReply() *WorkflowFutureReadyReply {
	reply := new(WorkflowFutureReadyReply)
	reply.WorkflowReply = NewWorkflowReply()
	reply.SetType(internal.WorkflowFutureReadyReply)

	return reply
}

// -------------------------------------------------------------------------
// IProxyMessage interface methods for implementing the IProxyMessage interface

// Clone inherits docs from WorkflowReply.Clone()
func (reply *WorkflowFutureReadyReply) Clone() IProxyMessage {
	workflowFutureReadyReply := NewWorkflowFutureReadyReply()
	var messageClone IProxyMessage = workflowFutureReadyReply
	reply.CopyTo(messageClone)

	return messageClone
}

// CopyTo inherits docs from WorkflowReply.CopyTo()
func (reply *WorkflowFutureReadyReply) CopyTo(target IProxyMessage) {
	reply.WorkflowReply.CopyTo(target)
}

File: Lib/Neon.Cadence/Internal/WorkflowMessages/WorkflowFutureReadyReply.cs
//-----------------------------------------------------------------------------
// FILE:	    WorkflowFutureReadyReply.cs
// CONTRIBUTOR: Jeff Lill
// COPYRIGHT:	Copyright (c) 2005-2021 by neonFORGE LLC.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections.Generic;
using System.ComponentModel;

using Neon.Cadence;
using Neon.Common;

namespace Neon.Cadence.Internal
{
    /// <summary>
    /// <para>
    /// <b>proxy --> client:</b> This is a special reply message sent for workflow operations that
    /// are implemented in GOLANG as futures and may be executed in parallel.  <b>cadence-proxy</b>
    /// will send this message after it has submitted the operation to Cadence but before the future
    /// actually completes.  The .NET client uses this reply as an indication that another Cadence
    /// operation may be started.
    /// </para>
    /// <note>
    /// This message does not have a corresponding request message (which is why the name doesn't end with "Reply".
    /// </note>
    /// </summary>
    [InternalProxyMessage(InternalMessageTypes.WorkflowFutureReadyReply)]
    internal class WorkflowFutureReadyReply : WorkflowReply
    {
        /// <summary>
        /// Default constructor.
        /// </summary>
        public WorkflowFutureReadyReply()
        {
            Type = InternalMessageTypes.WorkflowFutureReadyReply;
        }

        /// <inheritdoc/>
        internal override ProxyMessage Clone()
        {
            var clone = new WorkflowFutureReadyReply();

            CopyTo(clone);

            return clone;
        }

        /// <inheritdoc/>
        protected override void CopyTo(ProxyMessage target)
        {
            base.CopyTo(target);
        }
    }
}
